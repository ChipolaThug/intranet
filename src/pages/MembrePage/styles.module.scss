.membre {
    padding: 0 15px;
}
