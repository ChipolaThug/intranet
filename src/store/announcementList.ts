import { PayloadAction, createSlice, createEntityAdapter } from "@reduxjs/toolkit"

import { StateRequest, toastError, elementListFetch } from "./utils"
import { Announcement } from "../services/announcement"
import { AppThunk, AppState } from "."
import { announcementListGet } from "../services/announcementAccessors"

const announcementAdapter = createEntityAdapter<Announcement>()

export const initialState = announcementAdapter.getInitialState({
    readyStatus: "idle",
} as StateRequest)

const announcementList = createSlice({
    name: "announcementList",
    initialState,
    reducers: {
        getRequesting: (state) => {
            state.readyStatus = "request"
        },
        getSuccess: (state, { payload }: PayloadAction<Announcement[]>) => {
            state.readyStatus = "success"
            announcementAdapter.setAll(state, payload)
        },
        getFailure: (state, { payload }: PayloadAction<string>) => {
            state.readyStatus = "failure"
            state.error = payload
        },
    },
})

export default announcementList.reducer
export const { getRequesting, getSuccess, getFailure } = announcementList.actions

export const fetchAnnouncementList = elementListFetch(
    announcementListGet,
    getRequesting,
    getSuccess,
    getFailure,
    (error: Error) => toastError(`Erreur lors du chargement des announcements: ${error.message}`)
)

const shouldFetchAnnouncementList = (state: AppState) =>
    state.announcementList.readyStatus !== "success"

export const fetchAnnouncementListIfNeed = (): AppThunk => (dispatch, getState) => {
    const { jwt } = getState().auth
    if (shouldFetchAnnouncementList(getState())) return dispatch(fetchAnnouncementList(jwt))

    return null
}
