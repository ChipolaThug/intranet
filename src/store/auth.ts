import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit"
import { AppState } from "."

// Define a type for the slice state
interface AuthState {
    jwt: string
    id: number
    roles: string[]
}

// Define the initial state using that type
const initialState: AuthState = {
    jwt: "",
    id: 0,
    roles: [],
}

export const auth = createSlice({
    name: "auth",
    initialState,
    reducers: {
        setCurrentUser: (state, action: PayloadAction<AuthState>) => {
            state.jwt = action.payload.jwt
            state.id = action.payload.id
            state.roles = action.payload.roles
        },
        logoutUser: (state) => {
            // Unused, just reload page :/
            state.jwt = ""
            state.id = 0
            state.roles = []
        },
    },
})

export const { setCurrentUser, logoutUser } = auth.actions

const selectAuthData = (state: AppState): AuthState => state.auth

const selectRouter = (state: AppState): AppState["router"] => state.router

export const selectUserJwtToken = createSelector(selectAuthData, (authData) => authData.jwt)

export const routerSelector = createSelector(selectRouter, (authData) => authData)

export const selectUserRoles = createSelector(selectAuthData, (authData) => authData.roles)

export const selectUserId = createSelector(selectAuthData, (authData) => authData.id)

export const isUserConnected = createSelector(selectUserJwtToken, (token) => !!token)

export default auth.reducer
