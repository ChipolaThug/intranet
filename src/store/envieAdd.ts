import { PayloadAction, createSlice, createEntityAdapter } from "@reduxjs/toolkit"
import { toast } from "react-toastify"

import { StateRequest } from "./utils"
import { Envie, EnvieWithoutId, addEnvie } from "../services/envies"
import { AppThunk } from "."

const envieAdapter = createEntityAdapter<Envie>({
    selectId: (envie) => envie.envieId,
})

const envieAdd = createSlice({
    name: "addEnvie",
    initialState: envieAdapter.getInitialState({
        readyStatus: "idle",
    } as StateRequest),
    reducers: {
        getRequesting: (state) => {
            state.readyStatus = "request"
        },
        getSuccess: (state, { payload }: PayloadAction<Envie>) => {
            state.readyStatus = "success"
            envieAdapter.addOne(state, payload)
        },
        getFailure: (state, { payload }: PayloadAction<string>) => {
            state.readyStatus = "failure"
            state.error = payload
        },
    },
})

export default envieAdd.reducer
export const { getRequesting, getSuccess, getFailure } = envieAdd.actions

export const postEnvie =
    (envieWithoutId: EnvieWithoutId): AppThunk =>
    async (dispatch) => {
        dispatch(getRequesting())

        const { error, data } = await addEnvie(envieWithoutId)

        if (error) {
            dispatch(getFailure(error.message))
            toast.error(`Erreur lors de l'ajout d'une envie: ${error.message}`, {
                position: "top-center",
                autoClose: 6000,
                hideProgressBar: true,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
            })
        } else {
            dispatch(getSuccess(data as Envie))
            toast.success("Envie ajout√©e !", {
                position: "top-center",
                autoClose: 3000,
                hideProgressBar: true,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
            })
        }
    }
