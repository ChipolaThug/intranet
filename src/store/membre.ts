import { PayloadAction, createSlice } from "@reduxjs/toolkit"
import { toast } from "react-toastify"

import { StateRequest } from "./utils"
import { Membre, getMembre } from "../services/membres"
import { AppThunk, AppState } from "."

type StateMembre = { entity?: Membre } & StateRequest

export const initialState: StateMembre = {
    readyStatus: "idle",
}

const membre = createSlice({
    name: "membre",
    initialState,
    reducers: {
        getRequesting: (_) => ({
            readyStatus: "request",
        }),
        getSuccess: (_, { payload }: PayloadAction<Membre>) => ({
            readyStatus: "success",
            entity: payload,
        }),
        getFailure: (_, { payload }: PayloadAction<string>) => ({
            readyStatus: "failure",
            error: payload,
        }),
    },
})

export default membre.reducer
export const { getRequesting, getSuccess, getFailure } = membre.actions

export const fetchMembreData =
    (id: number): AppThunk =>
    async (dispatch) => {
        dispatch(getRequesting())

        const { error, data } = await getMembre(id)

        if (error) {
            dispatch(getFailure(error.message))
            toast.error(`Erreur lors du chargement du membre ${id}: ${error.message}`, {
                position: "top-center",
                autoClose: 6000,
                hideProgressBar: true,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
            })
        } else {
            dispatch(getSuccess(data as Membre))
        }
    }

const shouldFetchMembreData = (state: AppState, id: number) =>
    state.membre.readyStatus !== "success" || (state.membre.entity && state.membre.entity.id !== id)

export const fetchMembreDataIfNeed =
    (id: number): AppThunk =>
    (dispatch, getState) => {
        if (shouldFetchMembreData(getState(), id)) return dispatch(fetchMembreData(id))

        return null
    }
