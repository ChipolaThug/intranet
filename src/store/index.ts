import { createMemoryHistory, createBrowserHistory } from "history"
import { Action, configureStore, EntityState } from "@reduxjs/toolkit"
import { ThunkAction } from "redux-thunk"
import { routerMiddleware } from "connected-react-router"
import Cookies from "js-cookie"

import createRootReducer from "./rootReducer"
import { StateRequest } from "./utils"
import { setCurrentUser, logoutUser } from "./auth"

interface Arg {
    initialState?: typeof window.__INITIAL_STATE__
    url?: string
    jwt?: string
    id?: number
}

// Use inferred return type for making correctly Redux types
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const createStore = ({ initialState, url, jwt, id }: Arg = {}) => {
    const history = __SERVER__
        ? createMemoryHistory({ initialEntries: [url || "/"] })
        : createBrowserHistory()
    const store = configureStore({
        preloadedState: initialState,
        reducer: createRootReducer(history),
        middleware: (getDefaultMiddleware) => [
            // Included default middlewares: https://redux-toolkit.js.org/api/getDefaultMiddleware#included-default-middleware
            ...getDefaultMiddleware(),
            routerMiddleware(history),
        ],
        devTools: __DEV__,
    })

    if (jwt && id) {
        store.dispatch(setCurrentUser({ jwt, id }))
    } else {
        store.dispatch(logoutUser())
    }

    return { store, history }
}

const storage: any = localStorage
const id = +(Cookies.get("id") || storage?.getItem("id"))
const jwt = Cookies.get("jwt") || storage?.getItem("jwt")

const { store } = createStore({ id, jwt })

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch

export type AppThunk = ThunkAction<void, AppState, unknown, Action<string>>

export type EntitiesRequest<T> = EntityState<T> & StateRequest

export type ValueRequest<T> = { value?: T } & StateRequest

export default createStore
