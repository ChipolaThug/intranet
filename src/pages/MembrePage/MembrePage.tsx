import { useEffect, memo } from "react"
import { RouteComponentProps } from "react-router-dom"
import { useDispatch, useSelector, shallowEqual } from "react-redux"
import { Helmet } from "react-helmet"

import { AppState, AppThunk } from "../../store"
import { fetchMembreDataIfNeed } from "../../store/membre"
import { MembreInfo, MembreSet } from "../../components"
import styles from "./styles.module.scss"

export type Props = RouteComponentProps<{ id: string }>

const MembrePage = ({ match }: Props): JSX.Element => {
    const { id: rawId } = match.params
    const id = +rawId
    const dispatch = useDispatch()
    const membre = useSelector((state: AppState) => state.membre, shallowEqual)

    useEffect(() => {
        dispatch(fetchMembreDataIfNeed(id))
    }, [dispatch, id])

    const renderInfo = () => {
        const membreInfo = membre

        if (!membreInfo || membreInfo.readyStatus === "request") return <p>Loading...</p>

        if (membreInfo.readyStatus === "failure" || !membreInfo.entity)
            return <p>Oops! Failed to load data.</p>

        return (
            <div>
                <MembreInfo item={membreInfo.entity} />
                <MembreSet dispatch={dispatch} membre={membreInfo.entity} />
            </div>
        )
    }

    return (
        <div className={styles.Membre}>
            <Helmet title="User Info" />
            {renderInfo()}
        </div>
    )
}

interface LoadDataArgs {
    params: { id: number }
}

export const loadData = ({ params }: LoadDataArgs): AppThunk[] => [fetchMembreDataIfNeed(params.id)]

export default memo(MembrePage)
