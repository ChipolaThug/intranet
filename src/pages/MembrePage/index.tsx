import loadable from "@loadable/component"

import { Loading, ErrorBoundary } from "../../components"
import { Props, loadData } from "./MembrePage"

const MembrePage = loadable(() => import("./MembrePage"), {
    fallback: <Loading />,
})

export default (props: Props): JSX.Element => (
    <ErrorBoundary>
        <MembrePage {...props} />
    </ErrorBoundary>
)
export { loadData }
