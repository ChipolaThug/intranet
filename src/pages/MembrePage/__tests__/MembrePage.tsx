/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"

import { fetchMembreIfNeed } from "../../../store/membre"
import mockStore from "../../../utils/mockStore"
import MembrePage from "../MembrePage"

describe("<MembrePage />", () => {
    const mockData = {
        id: 1,
        nom: "Aupeix",
        prenom: "Amélie",
        mail: "pakouille.lakouille@yahoo.fr",
        telephone: "0675650392",
        photo: "images/membres/$taille/amélie_aupeix.jpg",
        alimentation: "Végétarien",
        majeur: 1,
        privilege: 0,
        actif: 0,
        commentaire: "",
        horodatage: "0000-00-00",
        passe: "$2y$10$fSxY9AIuxSiEjwF.J3eXGubIxUPlobkyRrNIal8ASimSjNj4SR.9O",
    }
    const { id } = mockData

    const renderHelper = (reducer = {}) => {
        const { dispatch, ProviderWithStore } = mockStore({ membre: reducer })
        const { container } = render(
            <ProviderWithStore>
                <MemoryRouter>
                    {/*
            @ts-expect-error */}
                    <MembrePage match={{ params: { id } }} />
                </MemoryRouter>
            </ProviderWithStore>
        )

        return { dispatch, firstChild: container.firstChild }
    }

    it("should fetch data when page loaded", () => {
        const { dispatch } = renderHelper()

        expect(dispatch).toHaveBeenCalledTimes(1)
        expect(dispatch.mock.calls[0][0].toString()).toBe(fetchMembreIfNeed(id).toString())
    })

    it("renders the loading status if data invalid", () => {
        expect(renderHelper().firstChild).toMatchSnapshot()
    })

    it("renders the loading status if requesting data", () => {
        const reducer = { readyStatus: "request" }

        expect(renderHelper(reducer).firstChild).toMatchSnapshot()
    })

    it("renders an error if loading failed", () => {
        const reducer = { readyStatus: "failure" }

        expect(renderHelper(reducer).firstChild).toMatchSnapshot()
    })

    it("renders the <Info /> if loading was successful", () => {
        const reducer = { readyStatus: "success", entity: mockData }

        expect(renderHelper(reducer).firstChild).toMatchSnapshot()
    })
})
