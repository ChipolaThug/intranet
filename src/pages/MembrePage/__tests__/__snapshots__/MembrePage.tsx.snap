// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MembrePage /> renders an error if loading failed 1`] = `
<div
  class="membre"
>
  <p>
    Oops! Failed to load data.
  </p>
</div>
`;

exports[`<MembrePage /> renders the <Info /> if loading was successful 1`] = `
<div
  class="membre"
>
  <div>
    <div
      class="MembreCard"
    >
      <h4>
        Membre Info
      </h4>
      <ul>
        <li>
          Prénom: 
          Amélie
        </li>
        <li>
          Nom: 
          Aupeix
        </li>
      </ul>
    </div>
    <section
      class="MembreList"
    >
      <h2>
        Modifier un membre
      </h2>
      <form>
        <label
          for="postPrenom"
        >
          Prenom:
          <input
            id="postPrenom"
            name="postPrenom"
            type="text"
            value="Amélie"
          />
        </label>
        <label
          for="postNom"
        >
          Nom:
          <input
            id="postNom"
            name="postNom"
            type="text"
            value="Aupeix"
          />
        </label>
        <label
          for="postMajeur"
        >
          Majeur:
          <input
            id="postMajeur"
            name="postMajeur"
            type="text"
            value="1"
          />
        </label>
        <button
          type="button"
        >
          Save changes
        </button>
      </form>
    </section>
  </div>
</div>
`;

exports[`<MembrePage /> renders the loading status if data invalid 1`] = `
<div
  class="membre"
>
  <p>
    Oops! Failed to load data.
  </p>
</div>
`;

exports[`<MembrePage /> renders the loading status if requesting data 1`] = `
<div
  class="membre"
>
  <p>
    Loading...
  </p>
</div>
`;
